keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,350 @@
+	// which will determine the default filesystem to apply.
+	//
+	// +optional
+	FSType *string
+
+	// VolumeAttributes store immutable properties of the volume copied during provision.
+	// These attributes are passed back to the driver during controller publish calls.
+	// +optional
+	VolumeAttributes map[string]string
+
+	// ControllerPublishSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// ControllerPublishVolume and ControllerUnpublishVolume calls.
+	// This field is optional, and  may be empty if no secret is required. If the
+	// secret object contains more than one secret, all secret references are passed.
+	// 
+	// +optional
+	ControllerPublishSecretRef *LocalObjectReference
+
+	// NodeStageSecretRef is a reference to the secret object containing sensitive
+	// information to pass to the CSI driver to complete the CSI NodeStageVolume
+	// and NodeStageVolume and NodeUnstageVolume calls.
+	// This field is optional, and  may be empty if no secret is required. If the
+	// secret object contains more than one secret, all secret references are passed.
+	// 
+	// +optional
+	NodeStageSecretRef *LocalObjectReference
+
+	// NodePublishSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// NodePublishVolume and NodeUnpublishVolume calls.
+	// This field is optional, and  may be empty if no secret is required. If the
+	// secret object contains more than one secret, all secret references are passed.
+	// 
+	// +optional
+	NodePublishSecretRef *LocalObjectReference
+}
+```
+
+### Secret references
+All secret references declared in a CSI inline volume can only refer to pre-defined secrets within the same namespace where the corresponding pod is running.  Secret references are stored in `LocalObjectReference` instances which require a pointer to the secret and a namespace value.
+
+### Specifying allowed inline drivers with `PodSecurityPolicy`
+To control which driver is allowed to be used within a pod spec, this design will update the `PodSecurityPolicy` to introduce `AllowedCSIDrivers` as shown below:
+
+
+```go
+  type PodSecurityPolicySpec struct {
+	// <snip>
+
+	// AllowedCSIDrivers is a whitelist of allowed CSI drivers.  Empty or nil indicates that all
+	// CSI drivers may be used.  This parameter is effective only when the usage of the CSI plugin
+	// is allowed in the "Volumes" field.
+	// +optional
+	AllowedCSIDrivers []AllowedCSIDriver
+  }
+
+  // AllowedCSIDriver represents a single CSI driver that is allowed to be used.
+  type AllowedCSIDriver struct {
+	// Name of the CSI driver
+	Name string
+  }  
+```
+
+Value `PodSecurityPolicy.AllowedCSIDrivers` must be explicitly set with the names of CSI drivers that are allowed to be embedded within a pod spec.  An empty value means no CSI drivers are allowed to be specified inline inside a pod spec.
```

Comment:
Can we make CSI drivers for ephemeral volumes allowed by default? Or will that break something? It would be pretty crappy if I had to set this PodSecuirtyPolicy policy to use, for example, secret volumes?

Reply:
its consistent with what already exists in PSP. if you want to use a configmap or secret volume, you have to list them explicitly. for example, here's what I have on one of my clusters:
```yaml
  volumes:
  - configMap
  - emptyDir
  - projected
  - secret
  - downwardAPI
  - persistentVolumeClaim
```

Reply:
flexVolumes in the psp also work the same way. a white list of allowed drivers for that psp.

Reply:
Also, keep in mind that inline CSI drivers will not necessarily replace built-in secrets, configmap, emptyDir, downwardAPI drivers.

Reply:
I forgot, PSP is not enforced if the PSP object doesn't exist. So by default there is no enforcement. If you create a PSP, only then there is enforcement. Which satisfies my use case.

Reply:
> I forgot, PSP is not enforced if the PSP object doesn't exist. So by default there is no enforcement. If you create a PSP, only then there is enforcement.

That is not accurate... enablement of the PSP admission plugin enables enforcement. Once enforcement is enabled, it is deny by default. If no PSPs exist, no pods are allowed.

Reply:
With the introduction of a new volume type, clusters with existing PSPs should consider whether to enable it or not. If they choose not to enable it, use of that new volume type would not be permitted.

Reply:
yeah. I think he meant, it would be allowed automatically if you hadn't enabled psp's?

Reply:
I see thanks for the clarification. So this means that if a Kubernetes distributor ever decided to go from PSP admission plugin disabled to enabled by default, the distributor would also need to either figure out all the drivers to whitelist or they would break users? Seems like that is a problem outside of the scope of this implementation.